File: infra/image/formats/gemini-native-image.ts
/**
 * Gemini Native API 이미지 생성 핸들러 — gemini-2.5-flash-image (Nano Banana).
 *
 * 요청: POST https://generativelanguage.googleapis.com/v1beta/models/<model>:generateContent
 * 인증: ?key=<API_KEY> query param (또는 x-goog-api-key 헤더)
 *
 * Request body:
 *   { contents: [{ parts: [{ text: prompt }] }] }
 *
 * Response:
 *   { candidates: [{ content: { parts: [{ inline_data: { mime_type, data (base64) } }] } }] }
 *
 * Gemini 2.5 Flash Image 특징:
 *  - 사이즈는 모델이 프롬프트 기반 자동 판단 (size 파라미터 없음)
 *  - 품질도 고정 (standard)
 *  - 한국어 텍스트 렌더링 우수
 *  - 캐릭터·객체 일관성 강점
 */
import type { ImageGenOpts, ImageGenCallOpts, ImageGenResult } from '../../../core/ports';
import type { InfraResult } from '../../../core/types';
import type { ImageFormatHandler, ImageFormatHandlerContext } from '../format-handler';

export class GeminiNativeImageFormat implements ImageFormatHandler {
  async generate(
    opts: ImageGenOpts,
    _callOpts: ImageGenCallOpts | undefined,
    ctx: ImageFormatHandlerContext,
  ): Promise<InfraResult<ImageGenResult>> {
    const apiKey = ctx.resolveApiKey();
    if (!apiKey) return { success: false, error: `API 키가 설정되지 않았습니다: ${ctx.config.apiKeyVaultKey}` };

    // Gemini 는 size 를 직접 받지 않음. 프롬프트에 aspect ratio 힌트 주입 (사용자 의도 보존).
    let prompt = opts.prompt;
    if (opts.size && opts.size !== 'auto') {
      const hint = opts.size.replace('x', ':');
      prompt = `${prompt}\n\n(Aspect ratio hint: ${hint}.)`;
    }

    // referenceImage (image-to-image) — Gemini 2.5+ Flash Image 는 multimodal contents 자연 지원.
    // parts 에 inline_data (image) + text 함께 보내면 image-to-image 변환 동작.
    // text 가 변환 의도 (스타일 변경 / 실사화 등) 설명하면 됨.
    const ref = opts.referenceImage;
    const parts: Array<{ text?: string } | { inline_data?: { mime_type: string; data: string } }> = [];
    if (ref) {
      parts.push({ inline_data: { mime_type: ref.contentType, data: ref.binary.toString('base64') } });
    }
    parts.push({ text: prompt });

    const url = `${ctx.config.endpoint}?key=${encodeURIComponent(apiKey)}`;
    const body = {
      contents: [{ parts }],
      generationConfig: {
        responseModalities: ['IMAGE'],
      },
    };

    try {
      const res = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      });
      if (!res.ok) {
        const txt = await res.text().catch(() => '');
        const flat = txt.replace(/\s+/g, ' ').slice(0, 2000);
        return { success: false, error: `Gemini Images API ${res.status}: ${flat}` };
      }
      const json = await res.json() as {
        candidates?: Array<{
          content?: {
            parts?: Array<{
              inline_data?: { mime_type?: string; data?: string };
              inlineData?: { mimeType?: string; data?: string };
            }>;
          };
        }>;
      };
      const parts = json.candidates?.[0]?.content?.parts ?? [];
      // Gemini 응답은 snake_case (inline_data) 또는 camelCase (inlineData) — 둘 다 허용
      const imagePart = parts.find(p => p.inline_data?.data || p.inlineData?.data);
      const data = imagePart?.inline_data?.data ?? imagePart?.inlineData?.data;
      const mimeType = imagePart?.inline_data?.mime_type ?? imagePart?.inlineData?.mimeType ?? 'image/png';
      if (!data) return { success: false, error: '응답에 이미지 데이터가 없습니다' };
      const binary = Buffer.from(data, 'base64');
      return {
        success: true,
        data: { binary, contentType: mimeType },
      };
    } catch (err: unknown) {
      const msg = err instanceof Error ? err.message : String(err);
      return { success: false, error: msg };
    }
  }
}

File: core-rust/src/image_gen/formats/gemini_native_image.rs
//! Gemini Native API 이미지 생성 — gemini-3.1-flash-image-preview (Nano Banana 후속).
//!
//! 옛 TS `infra/image/formats/gemini-native-image.ts` 1:1 port.
//! POST `https://generativelanguage.googleapis.com/v1beta/models/<model>:generateContent`
//! 인증: `?key=<API_KEY>` query param.
//! Multimodal contents — reference_image 박혀있으면 inline_data part 자동 추가.

use base64::Engine;
use serde::Deserialize;

use crate::image_gen::config::compute_image_cost;
use crate::image_gen::format_handler::{ImageFormatHandler, ImageFormatHandlerContext};
use crate::ports::{ImageGenCallOpts, ImageGenOpts, ImageGenResult, InfraResult};

#[derive(Debug, Deserialize)]
struct GeminiResponse {
    candidates: Option<Vec<GeminiCandidate>>,
}

#[derive(Debug, Deserialize)]
struct GeminiCandidate {
    content: Option<GeminiContent>,
}

#[derive(Debug, Deserialize)]
struct GeminiContent {
    parts: Option<Vec<GeminiPart>>,
}

/// Gemini 응답은 snake_case (`inline_data`) 또는 camelCase (`inlineData`) — 둘 다 허용.
#[derive(Debug, Deserialize)]
struct GeminiPart {
    #[serde(rename = "inline_data", default)]
    inline_data: Option<GeminiInlineData>,
    #[serde(rename = "inlineData", default)]
    inline_data_camel: Option<GeminiInlineData>,
}

#[derive(Debug, Deserialize)]
struct GeminiInlineData {
    #[serde(rename = "mime_type", default)]
    mime_type_snake: Option<String>,
    #[serde(rename = "mimeType", default)]
    mime_type_camel: Option<String>,
    data: Option<String>,
}

impl GeminiInlineData {
    fn mime_type(&self) -> String {
        self.mime_type_snake
            .clone()
            .or_else(|| self.mime_type_camel.clone())
            .unwrap_or_else(|| "image/png".to_string())
    }
}

pub struct GeminiNativeImageFormat;

impl GeminiNativeImageFormat {
    pub fn new() -> Self {
        Self
    }
}

impl Default for GeminiNativeImageFormat {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait::async_trait]
impl ImageFormatHandler for GeminiNativeImageFormat {
    async fn generate(
        &self,
        opts: &ImageGenOpts,
        _call_opts: &ImageGenCallOpts,
        ctx: ImageFormatHandlerContext<'_>,
    ) -> InfraResult<ImageGenResult> {
        let api_key = (ctx.resolve_api_key)().ok_or_else(|| {
            format!(
                "API 키가 설정되지 않았습니다: {}",
                ctx.config.api_key_vault_key
            )
        })?;

        // Gemini 는 size 직접 X — 프롬프트에 aspect ratio 힌트 주입 (사용자 의도 보존).
        let mut prompt = opts.prompt.clone();
        if let Some(size) = opts.size.as_deref() {
            if size != "auto" {
                let hint = size.replace('x', ":");
                prompt.push_str(&format!("\n\n(Aspect ratio hint: {}.)", hint));
            }
        }

        // multimodal contents.parts — reference_image 박혀있으면 inline_data 먼저, 그 다음 text.
        let mut parts: Vec<serde_json::Value> = Vec::new();
        if let Some(ref_img) = &opts.reference_image {
            let b64 = base64::engine::general_purpose::STANDARD.encode(&ref_img.binary);
            parts.push(serde_json::json!({
                "inline_data": {
                    "mime_type": ref_img.content_type,
                    "data": b64,
                }
            }));
        }
        parts.push(serde_json::json!({ "text": prompt }));

        let url = format!(
            "{}?key={}",
            ctx.config.endpoint,
            urlencoding::encode(&api_key)
        );
        let body = serde_json::json!({
            "contents": [{ "parts": parts }],
            "generationConfig": {
                "responseModalities": ["IMAGE"],
            },
        });

        let client = crate::llm::formats::common::http_client();
        let resp = client
            .post(&url)
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("Gemini Images API 요청: {e}"))?;

        let status = resp.status();
        if !status.is_success() {
            let txt = resp.text().await.unwrap_or_default();
            let flat: String = txt.split_whitespace().collect::<Vec<_>>().join(" ");
            let truncated: String = flat.chars().take(2000).collect();
            return Err(format!("Gemini Images API {}: {}", status, truncated));
        }

        let json: GeminiResponse = resp
            .json()
            .await
            .map_err(|e| format!("Gemini 응답 JSON: {e}"))?;
        let first_part = json
            .candidates
            .and_then(|mut c| c.drain(..).next())
            .and_then(|c| c.content)
            .and_then(|c| c.parts)
            .and_then(|mut p| {
                // inline_data 가 박힌 part 우선 (text part 가 먼저 오는 경우 있음)
                p.drain(..).find(|part| {
                    part.inline_data
                        .as_ref()
                        .and_then(|d| d.data.as_ref())
                        .is_some()
                        || part
                            .inline_data_camel
                            .as_ref()
                            .and_then(|d| d.data.as_ref())
                            .is_some()
                })
            })
            .ok_or_else(|| "응답에 이미지 데이터가 없습니다".to_string())?;
        let inline_data = first_part
            .inline_data
            .or(first_part.inline_data_camel)
            .ok_or_else(|| "응답 part 의 inline_data 비어있음".to_string())?;
        let mime_type = inline_data.mime_type();
        let b64 = inline_data
            .data
            .ok_or_else(|| "응답 inline_data.data 비어있음".to_string())?;
        let binary = base64::engine::general_purpose::STANDARD
            .decode(b64)
            .map_err(|e| format!("base64 decode: {e}"))?;

        // Gemini 는 quality 무관 단일 단가 (config.pricing.perImage)
        let cost_usd = compute_image_cost(ctx.config, opts.quality.as_deref());

        Ok(ImageGenResult {
            binary,
            content_type: mime_type,
            width: None,
            height: None,
            revised_prompt: None,
            cost_usd,
        })
    }
}

// urlencoding crate 의존성 회피 — 단순 percent encode 헬퍼 재사용 (LLM 의 Gemini handler 와 동일 패턴).
mod urlencoding {
    pub fn encode(s: &str) -> String {
        s.chars()
            .map(|c| match c {
                'a'..='z' | 'A'..='Z' | '0'..='9' | '-' | '_' | '.' | '~' => c.to_string(),
                _ => format!("%{:02X}", c as u8),
            })
            .collect()
    }
}
